c++/src/capnp/arena.h

```diff
@@ -91,11 +99,35 @@ class ReadLimiter {
 
 private:
   volatile uint64_t limit;
-  // Current limit, decremented each time catRead() is called.  Volatile because multiple threads
-  // could be trying to modify it at once.  (This is not real thread-safety, but good enough for
-  // the purpose of this class.  See class comment.)
+  // Current limit, decremented each time catRead() is called. We modify this variable using atomics
+  // with "relaxed" thread safety to make TSAN happy. See the class comment for why this is OK (
+  // previously we used a volatile variable).
 
   KJ_DISALLOW_COPY(ReadLimiter);
+
+  KJ_ALWAYS_INLINE(void setLimit(uint64_t newLimit)) {
+#if _MSC_VER && !defined(__clang__)
+#if _M_ARM
+    _InterlockedExchange64_nf(reinterpret_cast<volatile int64_t*>(&limit), newLimit);
+#else
+    _InterlockedExchange64(reinterpret_cast<volatile int64_t*>(&limit), newLimit);
+#endif
+#else
+    __atomic_store_n(&limit, newLimit, __ATOMIC_RELAXED);
+#endif
+  }
+
+  KJ_ALWAYS_INLINE(uint64_t readLimit() const) {
+#if _MSC_VER && !defined(__clang__)
+#if _M_ARM
+    return _InterlockedOr64_nf(reinterpret_cast<volatile int64_t*>(&limit), 0);
+#else
+    return _InterlockedOr64(reinterpret_cast<volatile int64_t*>(&limit), 0);
```

Comment:
Similarly, `_InterlockedOr64` actually stores the result back to the target variable, making this a much more expensive atomic operation than a simple load.